## **Análise de Commits em um Repositório Git**

- **1° Passo:**
  - **Descrição:** Primeiramente, foi utilizado o comando <span style="color:blue">`git log --pretty=format:"%H;%ae" > commitsAndroid.csv`</span> , que gera um arquivo CSV contendo os hashes de todos os commits e os e-mails dos desenvolvedores responsáveis por cada commit.


- **2° Passo:**
  - **Descrição:** Em seguida, utilizou-se o comando <span style="color:blue">`git rev-list --count HEAD`</span>para verificar se o número total de commits correspondia ao número de linhas do arquivo CSV. Essa verificação assegura que todos os commits realizados no repositório foram capturados adequadamente. formate esse texto na formatação do texto acima.


- **3° Passo:**
  - **Descrição:** Em analises anteriores diversos problemas com a codificação do texto foram enfrentados, ja que a biblioteca pandas utiliza por padão a codificação UTF-8, para que esse problemas fossem contornado utilizamos a biblioteca chardet para encontrar qual a codificação do nosso csv gerado anteriormente.

In [ ]:
import chardet

with open(r'C:\\Users\\enzov\Desktop\\IC JOHNATAN\\ESTUDO GITHUB\\Android-MaterialRefreshLayout\\commitsAndroid.csv', 'rb') as f:
    rawdata = f.read()
    result = chardet.detect(rawdata)
    print(result)

{'encoding': 'UTF-16', 'confidence': 1.0, 'language': ''}


- **4° Passo:**
  - **Descrição:** Após a identificação da codificação o formato do csv foi alterado para o formato xlsx(Excel) para que as manipulações da tabela utilizando a biblioteca Pandas ficassem mais simples e menos erros ocorressem.

In [1]:
import pandas as pd

csvAndroid = pd.read_csv(r'C:\\Users\\enzov\\Desktop\\IC JOHNATAN\\ESTUDO GITHUB\\Android-MaterialRefreshLayout\\commitsAndroid.csv', encoding='UTF-16')

csvAndroid.columns = ['Hash', 'Email']

csvAndroid.to_excel('comitsAndroidExcel.xlsx', index=False)

- **5° Passo:**
  - **Descrição:** Nessa parte vamos começar a pegar dados dos nossos hashes, a partir dos hashes vamos utilizar o comando<span style="color:blue">`git show --pretty= --name-only <hash_commit>`</span> que é usado para obter informações sobre um commit específico no repositório clonado no nosso caso o comando verifica quais arquivos foram alterados nos commits, após saber o hash do commit e os arquivos dele o comando utilizado é o <span style="color:blue">`git blame --line-porcelain <hash_commit> -- <arquivo>`</span> que exibe informações detalhadas sobre cada linha de um arquivo específico em um commit específico então o algoritimo busca pela palavra "import", assim que a palavra é encontrada o algoritimo pega a linha até o ";". Então após verificar cada um dos commits uma nova coluna "Imports" é adicionada a nossa palilha contendo os dados obtidos pelo algoritimo.

In [ ]:
import os
import subprocess
import pandas as pd
import re

repo_path = r"C:\\Users\\enzov\\git\Android-MaterialRefreshLayout"

os.chdir(repo_path)

input_excel = r'C:\\Users\\enzov\\Desktop\\IC JOHNATAN\\ESTUDO GITHUB\\Android-MaterialRefreshLayout\\comitsAndroidExcel.xlsx'

data = pd.read_excel(input_excel)

import_regex = r"import\s+([\w\.]+)"

resultados = []

for index, row in data.iterrows():
    hash_commit = row['Hash']
    try:
        comando_files = ["git", "show", "--pretty=", "--name-only", hash_commit]
        arquivos = subprocess.check_output(comando_files, text=True).splitlines()
        
        imports = set()

        for arquivo in arquivos:
            if arquivo.endswith(".java"):
                comando_blame = ["git", "blame", "--line-porcelain", hash_commit, "--", arquivo]
                output_blame = subprocess.check_output(comando_blame, text=True, encoding="utf-8")

                for linha in output_blame.splitlines():
                    if "import" in linha:
                        matches = re.findall(import_regex, linha)
                        imports.update(matches)

        resultados.append("; ".join(imports) if imports else "")

    except subprocess.CalledProcessError as e:
        print(f"Erro ao processar commit {hash_commit}: {e}")
        resultados.append("")
    except UnicodeDecodeError as e:
        print(f"Erro de decodificação no commit {hash_commit}: {e}")
        resultados.append("")

data['Imports'] = resultados

data.to_excel(input_excel, index=False)

print(f"Arquivo atualizado com imports salvo em: {input_excel}")


Erro ao processar commit 4b4a62f63dba8d4e493387b7d418cc8cc5a6fb5a: Command '['git', 'blame', '--line-porcelain', '4b4a62f63dba8d4e493387b7d418cc8cc5a6fb5a', '--', 'app/src/main/java/com/cjj/header/SunFaceView.java']' returned non-zero exit status 128.
Erro ao processar commit b4e7a160b5df0ad27dbdb655172e63a8deda259d: Command '['git', 'blame', '--line-porcelain', 'b4e7a160b5df0ad27dbdb655172e63a8deda259d', '--', 'Library/src/androidTest/java/com/cjj/ApplicationTest.java']' returned non-zero exit status 128.
Erro ao processar commit d3732d0ba282e986c91516f13088e9e4922d3370: Command '['git', 'blame', '--line-porcelain', 'd3732d0ba282e986c91516f13088e9e4922d3370', '--', 'app/src/main/java/com/cjj/android_materialrefreshlayout/MainActivity.java']' returned non-zero exit status 128.
Arquivo atualizado com imports salvo em: C:\\Users\\enzov\\Desktop\\IC JOHNATAN\\ESTUDO GITHUB\\Android-MaterialRefreshLayout\\comitsAndroidExcel.xlsx


- **6° Passo:**
 - **Descrição:** Este código tem como objetivo processar um arquivo Excel contendo uma coluna chamada "Imports", onde cada célula pode ter uma string com várias importações de pacotes que são separadas por ";". O código lê o arquivo Excel usando a função pd.read_excel() e aplica uma função personalizada à coluna "Imports". Essa função, contar_pontos_virgula(), conta o número de ocorrências do caractere ; em cada string da coluna "Imports". A contagem é incrementada em 1 para cada ocorrência do ;, e o resultado é armazenado em uma nova coluna chamada "Contagem_Imports" e da o numero de imports que cada commiter fez de acordo com o hash do commit. Após o processamento, o arquivo Excel é sobrescrito com a nova coluna, e o arquivo atualizado é salvo no mesmo local original.

In [16]:
import pandas as pd

file_path = r'C:\\Users\\enzov\\Desktop\\IC JOHNATAN\\ESTUDO GITHUB\\Android-MaterialRefreshLayout\\comitsAndroidExcel.xlsx'

data = pd.read_excel(file_path)

def contar_pontos_virgula(imports_string):
    if isinstance(imports_string, str):  
        return imports_string.count(";") + 1
    return 0 

data['Contagem_Imports'] = data['Imports'].apply(contar_pontos_virgula)

data.to_excel(file_path, index=False)

print(f"Arquivo atualizado e salvo em: {file_path}")


Arquivo atualizado e salvo em: C:\\Users\\enzov\\Desktop\\IC JOHNATAN\\ESTUDO GITHUB\\Android-MaterialRefreshLayout\\comitsAndroidExcel.xlsx


- **7° Passo:**
 - **Descrição:** O código tem como objetivo processar os dados de um arquivo Excel que contém hashes de commits de um repositório Git. Para cada hash de commit, ele executa o comando<span style="color:blue">`git show --numstat <hash_commit>`</span> para obter informações sobre as linhas de código adicionadas e removidas no commit. Em seguida, o código utiliza uma expressão regular para garantir que apenas as linhas de código (não comentadas) sejam consideradas. As contagens de linhas de código adicionadas e removidas são então adicionadas ao DataFrame como novas colunas: "Linhas Adicionadas (Código)" e "Linhas Removidas (Código)". Por fim, o DataFrame atualizado é salvo no arquivo Excel.

In [17]:
import os
import subprocess
import pandas as pd
import re

repo_path = r"C:\\Users\\enzov\\git\\Android-MaterialRefreshLayout"

os.chdir(repo_path)

data = pd.read_excel(r'C:\\Users\\enzov\\Desktop\\IC JOHNATAN\\ESTUDO GITHUB\\Android-MaterialRefreshLayout\\comitsAndroidExcel.xlsx')

code_line_regex = r"^\s*(?!//)(?!/\*)(?!\*/)(.+?)(?<!\\)$"

data['Linhas Adicionadas (Código)'] = 0
data['Linhas Removidas (Código)'] = 0

for i, hash_commit in enumerate(data['Hash']):
    comando = ["git", "show", "--numstat", hash_commit]
    try:
        output = subprocess.check_output(comando, text=True)
        linhas_adicionadas_codigo = 0
        linhas_removidas_codigo = 0

        for linha in output.splitlines():
            if re.match(r'^\d+\s+\d+\s+', linha):  
                partes = linha.split("\t")
                adicionadas = int(partes[0])
                removidas = int(partes[1])

                if adicionadas > 0 and any(re.match(code_line_regex, linha) for linha in output.splitlines()):
                    linhas_adicionadas_codigo += adicionadas
                elif removidas > 0 and any(re.match(code_line_regex, linha) for linha in output.splitlines()):
                    linhas_removidas_codigo += removidas

        data.at[i, 'Linhas Adicionadas (Código)'] = linhas_adicionadas_codigo
        data.at[i, 'Linhas Removidas (Código)'] = linhas_removidas_codigo

    except subprocess.CalledProcessError:
        print(f"Erro ao processar commit: {hash_commit}")
    except UnicodeDecodeError as e:
        print(f"Erro de decodificação no commit {hash_commit}: {e}")

output_path = r"C:\\Users\\enzov\\Desktop\\IC JOHNATAN\\ESTUDO GITHUB\\Android-MaterialRefreshLayout\\comitsAndroidExcel.xlsx"

data.to_excel(output_path, index=False)

print(f"Arquivo salvo em: {output_path}")


Arquivo salvo em: C:\\Users\\enzov\\Desktop\\IC JOHNATAN\\ESTUDO GITHUB\\Android-MaterialRefreshLayout\\comitsAndroidExcel.xlsx


- **8° Passo:**
 - **Descrição:** O Proximo codigo busca identificar cada um dos desenvolvedores presentes no código e qual o numero de commits de cada um.

In [18]:
file_path = r'C:\\Users\\enzov\\Desktop\\IC JOHNATAN\\ESTUDO GITHUB\\Android-MaterialRefreshLayout\\comitsAndroidExcel.xlsx'

data = pd.read_excel(file_path)

CommitersUnicos = data['Email'].unique()
print(CommitersUnicos)

print("\nQuantidade de commits que cada um fez: ")
ContagemCommitersUnicos = data['Email'].value_counts()
print(ContagemCommitersUnicos)

['929178101@qq.com' 'qsh@yomnn.cn' 'pepa.amorim@gmail.com'
 'liaohuqiu@gmail.com']

Quantidade de commits que cada um fez: 
Email
929178101@qq.com         85
pepa.amorim@gmail.com     2
qsh@yomnn.cn              1
liaohuqiu@gmail.com       1
Name: count, dtype: int64


- **9° Passo:**
 - **Descrição:** Por fim buscamos quantas linhas de código cada um dos desenvolvedores adicionou no código ao longo de todos os commits ja feitos.

In [ ]:
ContLinhasDecodigo = data.groupby('Email')['Linhas Adicionadas (Código)'].sum().reset_index()

print(ContLinhasDecodigo)

                   Email  Linhas Adicionadas (Código)
0       929178101@qq.com                        12063
1    liaohuqiu@gmail.com                            2
2  pepa.amorim@gmail.com                          134
3           qsh@yomnn.cn                            6


- **10° Passo:**
 - **Descrição:** Separa a planilha principal e cria uma nova planilha para cada um dos desenvolvedores presentes, as novas planilhas apenas as linhas com commits que possuiam imports, após a criação de cada uma pega toda a coluna imports e cria uma nova planilha com os imports unicos de cada desenvolvedor.

In [ ]:
import pandas as pd
import os

input_excel = r'C:\\Users\\enzov\\Desktop\\IC JOHNATAN\\ESTUDO GITHUB\\Android-MaterialRefreshLayout\\comitsAndroidExcel.xlsx'
output_directory = r'C:\\Users\\enzov\\Desktop\\IC JOHNATAN\\ESTUDO GITHUB\\Android-MaterialRefreshLayout\\Comiters'

data = pd.read_excel(input_excel)

os.makedirs(output_directory, exist_ok=True)

unique_comiters = data['Email'].dropna().unique()

for email in unique_comiters:

    comiter_data = data[(data['Email'] == email) & data['Imports'].notna()]
    
    if not comiter_data.empty:
        comiter_filtered_data = comiter_data[['Email', 'Imports']].copy()

        all_imports = []
        for imports in comiter_filtered_data['Imports']:
            all_imports.extend(imports.split(';')) 

        unique_imports = sorted(set([imp.strip() for imp in all_imports if imp.strip()]))  

        comiter_filtered_data.loc[:, 'Imports Únicos'] = ""  
        comiter_filtered_data.loc[0, 'Imports Únicos'] = "; ".join(unique_imports) 

        comiter_name = email.split('@')[0] 
        output_excel = os.path.join(output_directory, f"{comiter_name}.xlsx")

        comiter_filtered_data.to_excel(output_excel, index=False)

        print(f"Arquivo salvo para {email}: {output_excel}")

print("Processo concluído!")


Arquivo salvo para 929178101@qq.com: C:\\Users\\enzov\\Desktop\\IC JOHNATAN\\ESTUDO GITHUB\\Android-MaterialRefreshLayout\\Comiters\929178101.xlsx
Arquivo salvo para qsh@yomnn.cn: C:\\Users\\enzov\\Desktop\\IC JOHNATAN\\ESTUDO GITHUB\\Android-MaterialRefreshLayout\\Comiters\qsh.xlsx
Arquivo salvo para pepa.amorim@gmail.com: C:\\Users\\enzov\\Desktop\\IC JOHNATAN\\ESTUDO GITHUB\\Android-MaterialRefreshLayout\\Comiters\pepa.amorim.xlsx
Processo concluído!


- **11° Passo:**
 - **Descrição:** Quebra conteudo da coluna 'imports unicos' em uma string após isso a API da IA Gemini do google é consumida e um prompt é elaborado para que a IA classifique cada um dos desenvolvedores através de seus imports, nos dando por fim a especialização, foco e as habilidades chave de cada um dos desenvolvedores.

In [ ]:
import os
import pandas as pd
import google.generativeai as genai  

API_KEY = 'AIzaSyBIeWE-sXr_RJgQFwCUUAUJXr9ECagZgx4'

genai.configure(api_key=API_KEY)
model = genai.GenerativeModel('gemini-pro')

pasta = r"C:\\Users\\enzov\\Desktop\\IC JOHNATAN\\ESTUDO GITHUB\\Android-MaterialRefreshLayout\\Comiters"

coluna_alvo = 'Imports Únicos'

resultados = []

for arquivo in os.listdir(pasta):
    if arquivo.endswith('.xlsx') or arquivo.endswith('.xls'):  
        caminho_arquivo = os.path.join(pasta, arquivo)
        
        try:
            df = pd.read_excel(caminho_arquivo)
    
            if coluna_alvo in df.columns:
                conteudo_coluna = df[coluna_alvo].dropna().tolist() 
                
                resultados.append({
                    "arquivo": arquivo,
                    "conteudo": conteudo_coluna
                })
                
                print(f"Processado: {arquivo}, {len(conteudo_coluna)} linhas capturadas.")
            else:
                print(f"A coluna '{coluna_alvo}' não foi encontrada no arquivo: {arquivo}")
        
        except Exception as e:
            print(f"Erro ao processar o arquivo {arquivo}: {e}")

for resultado in resultados:

    print(f"\nArquivo: {resultado['arquivo']}")

    conteudo_como_string = "\n".join(resultado['conteudo'])

    response = model.generate_content(
        f"""
    Analyze the following libraries and determine developer specialization, 
    providing a concise and categorized answer: (
    it is essential that in the specialization line the developer is classified as, Front-end Developer, Back-end Developer, Full-stack Developer, Data Scientist, Data Engineer, Machine Learning Engineer, Mobile Developer, DevOps Engineer, Cloud Engineer, Security Engineer, Application Security Engineer, Game Developer, Embedded Systems Developer:, AR/VR Developer or Blockchain Developer )

    * **Bibliotecas:**
    {conteudo_como_string}

    **Formato da resposta:**
    -Especialização: [especialização]
    -Foco: [foco principal]
    -Habilidades: [lista de habilidades-chave]
    """
    )
    print(response.text)


Processado: 929178101.xlsx, 1 linhas capturadas.
Processado: pepa.amorim.xlsx, 1 linhas capturadas.
Processado: qsh.xlsx, 1 linhas capturadas.

Arquivo: 929178101.xlsx
- Especialização: Mobile Developer 
- Foco: Desenvolvimento de aplicativos móveis Android 
- Habilidades: Interface do usuário Android, animações Android, Gerenciamento de layouts, Eventos de toque, Criação de widgets personalizados

Arquivo: pepa.amorim.xlsx
**Especialização:** Mobile Developer
**Foco:** Desenvolvimento de aplicativos móveis Android
**Habilidades:**
* Android SDK
* Java/Kotlin
* Desenvolvimento de interface de usuário (UI)
* Animações
* Manipulação de eventos
* Gerenciamento de layout
* Depuração e otimização de desempenho

Arquivo: qsh.xlsx
- Especialização: Mobile Developer
- Foco: Desenvolvimento de aplicativos Android
- Habilidades: Android SDK, Java, XML, Animações


Foram testadas 3 IAs como o mesmo prompt e dados rodados no código anterior para verificar se as respostas principalmente do topico Especialização eram iguais ou parecidos

# **Chat GPT:**

### **Desenvovedor - 929178101:**
- **Especialização:** Mobile Developer
- **Foco:** Android Application Development
- **Habilidades:** 
    Proficient in Android SDK components (e.g., android.animation, android.view, android.widget)
    UI/UX development using Android Views and Layouts (RecyclerView, LinearLayout, Toolbar)
    Animation handling (ObjectAnimator, ValueAnimator, AnimationSet)
    Application lifecycle management (Application, Context, Intent)
    Graphics and rendering (Canvas, Paint, RadialGradient)
    Event handling and motion (MotionEvent, ViewPropertyAnimatorUpdateListener)
    Integration with third-party libraries (com.cjj.MaterialRefreshLayout, com.nineoldandroids.animation)
    Testing frameworks (ApplicationTestCase)
    Logging and debugging (android.util.Log)
    Resource management (Resources, TypedArray)
    Custom components (Drawable, ShapeDrawable, OvalShape)

### **Desenvolvedor - pepa.amorim:** 
- **Especialização:** Mobile Developer  
- **Foco:** Android User Interface and Animation Development  
- **Habilidades:**  
  - UI design and layout management (`android.view.View`, `ViewGroup`, `FrameLayout`)  
  - Animation handling and transitions (`AnimatorSet`, `ObjectAnimator`, `ViewPropertyAnimatorCompat`)  
  - Custom attribute handling (`android.content.res.TypedArray`, `android.util.AttributeSet`)  
  - Compatibility features for older Android versions (`android.support.v4.view.ViewCompat`)  
  - Event and gesture handling (`MotionEvent`, `android.view.Gravity`)  
  - Debugging and logging (`android.util.Log`)  
  - Interpolator-based animations (`DecelerateInterpolator`)  
  - Handling lists and scrollable views (`android.widget.AbsListView`)  
  - Contextual resource management (`android.content.Context`) 

### **Desenvolvedor - qsh:**
- **Especialização:** Mobile Developer  
- **Foco:** Android Application UI and Animation Development  
- **Habilidades:**  
  - Animation handling (`AnimatorSet`, `ObjectAnimator`)  
  - UI development and layout management (`View`, `FrameLayout`)  
  - Compatibility features for older Android versions (`ViewCompat`)  
  - Custom attribute and resource handling (`AttributeSet`, `Context`)  
  - Event alignment and positioning (`Gravity`)  
  - Debugging and logging (`Log`)  

# **Copilot:**

### **Desenvovedor - 929178101:**
**Especialização:** Mobile Developer
**Foco:** Desenvolvimento de Aplicações Android
**Habilidades:**
- Animação e gráficos com `android.animation` e `android.graphics`
- Ciclo de vida da aplicação com `android.app.Application` e `android.os.Bundle`
- Manipulação de contextos e intenções com `android.content.Context` e `android.content.Intent`
- Criação e gerenciamento de layouts com `android.view` e `android.widget`
- Utilização de bibliotecas de suporte como `android.support` e `android.support.v7`
- Implementação de animações avançadas com `com.nineoldandroids.animation`
- Integração de layouts personalizados com `com.cjj`
- Conhecimento em diversos elementos da interface do usuário, como `android.support.v7.widget.RecyclerView` e `android.widget.ListView`
- Gerenciamento de recursos e atributos com `android.content.res.Resources` e `android.util.AttributeSet`
- Implementação de interpoladores e transformações de animação com `android.view.animation`

### **Desenvolvedor - pepa.amorim:**
**Especialização:** Mobile Developer
**Foco:** Desenvolvimento de Aplicações Android
**Habilidades:**
- Animação e gráficos com `android.animation.AnimatorSet` e `android.animation.ObjectAnimator`
- Manipulação de contextos e recursos com `android.content.Context` e `android.content.res.TypedArray`
- Trabalho com cores utilizando `android.graphics.Color`
- Compatibilidade de versões com `android.os.Build`
- Animação de propriedades e compatibilidade com `android.support.v4.view`
- Manutenção de atributos de visualização e logs com `android.util.AttributeSet` e `android.util.Log`
- Gerenciamento de layouts e eventos de toque com `android.view.ViewGroup`, `android.view.Gravity`, e `android.view.MotionEvent`
- Implementação de interpoladores e animações com `android.view.animation.DecelerateInterpolator`
- Manipulação de componentes de UI como `android.widget.AbsListView` e `android.widget.FrameLayout`

### **Desenvolvedor - qsh:**
**Especialização:** Mobile Developer
**Foco:** Desenvolvimento de Aplicações Android
**Habilidades:**
- Animação com `android.animation.AnimatorSet` e `android.animation.ObjectAnimator`
- Manipulação de contextos com `android.content.Context`
- Compatibilidade de versões e propriedades de visualização com `android.support.v4.view.ViewCompat`
- Definição de atributos e logs com `android.util.AttributeSet` e `android.util.Log`
- Gerenciamento de layouts e posicionamento com `android.view.View`, `android.view.Gravity` e `android.widget.FrameLayout`

# **Meta AI:**

### **Desenvovedor - 929178101:**
**Especialização:** Mobile Developer
**Foco:** Desenvolvimento Android
**Habilidades:**
- Android SDK
- Java
- Kotlin (opcional)
- Desenvolvimento
- Animações e transições
- Trabalho com layouts e views
- Uso de RecyclerView e ListView
- Manipulação de imagens e gráficos
- Ferramentas e Bibliotecas
- Android Support Library
- NineOldAndroids (biblioteca de animação)
- MaterialRefreshLayout (biblioteca de refresh)

### **Desenvolvedor - pepa.amorim:**
**Especialização:** Mobile Developer
**Foco:** Desenvolvimento Android
**Habilidades:**
- Android SDK
- Java ou Kotlin
- Desenvolvimento de interfaces gráficas
- Animações e transições
- Trabalho com layouts e views
- Manipulação de eventos de toque (MotionEvent)
- Uso de recursos de visualização (View, ViewGroup)
- Android Support Library
- API de animação Android

### **Desenvolvedor - qsh:**
**Especialização:** Mobile Developer
**Foco:** Desenvolvimento Android
**Habilidades:**
- Android SDK
- Java ou Kotlin
- Desenvolvimento de interfaces gráficas
- Animações e transições
- Trabalho com layouts e views
- Manipulação de componentes visuais
- Ferramentas e Bibliotecas
- Android Support Library
- API de animação Android